[Merged by Bors] - feat(algebra/algebra/basic): define `restrict_scalars.linear_equiv`

ocfnash opened the pull request:
Also updating some doc-strings.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/algebra/basic): define `restrict_scalars.linear_equiv`

Also updating some doc-strings.

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1433,15 +1441,29 @@ instance restrict_scalars.module_orig [semiring A] [add_comm_monoid M] [I : modu
+  map_smul' := λ a x, rfl,
+  inv_fun   := id,
+  left_inv  := λ x, rfl,
+  right_inv := λ y, rfl, }
```
```suggestion
linear_equiv.refl A M
```

ocfnash replied:
Nice!

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1404,20 +1404,28 @@ end linear_map
+    above, it is invisible functor available for free through typeclass inference.
+
+A standard example of when one might want to invoke 1 would be if one has a vector space over a
+field of characteristic zero and wishes to make use of the `ℚ`-algebra structure. -/
```
This doesn't seem like a compelling justification; doesn't typeclass inference find the `ℚ`-algebra structure automatically?

ocfnash replied:
I don't think it can. For example:
```lean
import algebra.algebra.basic

variables (K V : Type*) [field K] [char_zero K] [add_comm_group V] [module K V]

-- Fails
example : module ℚ V := infer_instance

-- Succeeds
example : module ℚ (restrict_scalars ℚ K V) := infer_instance
```

ocfnash replied:
When it attempts the failed inference, it has no idea about `K`.

eric-wieser replied:
Ah, I guess the `is_scalar_tower` solution here would be "just pass `module ℚ V` and `is_scalar_tower ℚ K V` as a typeclass argument", which lowers the chance of diamonds. In practice, all concrete `V`s should provide these instances anyway.

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1433,15 +1441,29 @@ instance restrict_scalars.module_orig [semiring A] [add_comm_monoid M] [I : modu
+namespace restrict_scalars
+
+/-- `restrict_scalars` is an equivalence of modules over the semiring `A`. -/
+@[nolint unused_arguments]
```
I think this lint warning is valid and should be fixed. You don't need `A` to be an `R`-algebra for this to make sense; `R` is completely irrelevant to this definition.

ocfnash replied:
OK, done.

ocfnash committed:
Suggestions from CR

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1404,20 +1404,28 @@ end linear_map
+Warning: use this type synonym judiciously! Often what one wants instead of `restrict_scalars` is
+actually the following setup:
+`[comm_semiring R] [semiring A] [algebra R A] [module R M] [module A M] [is_scalar_tower R A M]` (*)
+which is just the way of saying "let `M` be a module over the `R`-algebra `A`".
```
I think a more spelt-out example might be useful here.
````suggestion
Warning: use this type synonym judiciously!

Consider an example where we want to construct an `R`-linear map between `M` and `A`, given
```lean
variables (R A M N : Type*) [comm_semiring R] [semiring A] [algebra R A] [module A M]
```
With the assumptions above, we can't directly state this as we have no `module R M` structure.
`restrict_scalars` permits this to be written as:
```lean
-- a `R`-module structure on `M` is provided by `restrict_scalars` which is compatible
example : restrict_scalars R A M →ₗ[R] A := sorry
```
However, the preferred way is typically to just add this extra structure as an argument,
```lean
-- a `R`-module structure on `M` and proof of its compatibility is provided by the user
example [module R M] [is_scalar_tower R A M] : M →ₗ[R] A := sorry
```
as in practice concrete `M`s will provide both these typeclasses anyway.
````

ocfnash replied:
Yep I think explicit example is very helpful. Updated!

ocfnash committed:
Suggestion of CR: explicit example in doc string

eric-wieser commented on src/algebra/algebra/basic.lean:
````diff
@@ -1404,20 +1404,49 @@ end linear_map
+variables [comm_semiring R] [semiring A] [algebra R A] [add_comm_monoid M] [module A M]
+```
+With the assumptions above, we can't directly state this as we have no `module R M` structure but
+`restrict_scalars` permits this to be written as:
````
```suggestion
With the assumptions above we can't directly state our map as we have no `module R M` structure, but
`restrict_scalars` permits it to be written as:
```

ocfnash committed:
Update src/algebra/algebra/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1430,18 +1459,27 @@ instance [I : add_comm_group M] : add_comm_group (restrict_scalars R A M) := I
 
+namespace restrict_scalars
+
+/-- `restrict_scalars` is an equivalence of modules over the semiring `A`. -/
+def linear_equiv : restrict_scalars R A M ≃ₗ[A] M := linear_equiv.refl A M
```
I thought about adding this in the past, but concluded there was little point because it's just a convoluted way to spell `id`. Do you actually end up needing this?

ocfnash replied:
I don't strictly need it but I claim that the way I use it here https://github.com/leanprover-community/mathlib/pull/7808/files#diff-8a09a95c8c28517dfa1a667d83cfdc6dff8cfbe366abb3c0e6000dc87bea4176R150 justifies its existence.

jcommelin commented on src/algebra/algebra/basic.lean:
```diff
@@ -1404,20 +1404,49 @@ end linear_map
+    perhaps a better name would be `enrich_scalars`).
+ 2. The functor from the category of modules over the `R`-algebra `A` to the category of modules
+    over the (semi)ring `R`. This is the true restriction of scalars but when one uses the setup
+    in the second example above, it is invisible functor available for free through typeclass
```
```suggestion
    in the second example above, it is an invisible functor available for free through typeclass
```

jcommelin commented on src/algebra/algebra/basic.lean:
```diff
@@ -1404,20 +1404,49 @@ end linear_map
+Note that this means we almost always want to state definitions and lemmas in the language of
+`is_scalar_tower` rather than `restrict_scalars`.
+
+Mathematically, when `A` is an `R`-algebra, there are two different functors:
```
I am completely confused by this remark. I thought that `restrict_scalars` is exactly meant to be used for (2) in situations where we need a "construction" (as opposed to just adding an `is_scalar_tower` assumption.

I don't understand what (1) means in Lean anyway. We only have modules over semirings. I don't think Lean distinguishes between modules over the semiring `A` and modules over the `R`-algebra `A`.

ocfnash replied:
Somehow, despite my best efforts, I seem not to be able to communicate my point clearly.

I'll make one last attempt to find the right words but given that you are [not the first person](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/restrict_scalars.20vs.20enrich_scalars/near/241172703) who finds my words unclear, probably the thing to do is just to drop this remark entirely.

I claim the confusion stems from the fact that typeclass inference (in conjunction with mixins) makes a lot of forgetful functors invisible. To avoid this I'll try to make my point using informal mathematical language.

Thus let:

 - `𝓜(S)` be the category of modules over a semiring `S`,
 - `𝓡(A, R)` be the category of representations of an `R`-algebra `A` (over `R`),
 - `F` be the functor that turns an algebra into a semiring (by forgetting the scalar action),
 - `G : 𝓡(A, R) ⥤ 𝓜(R)` be the functor that forgets the A action.

Note there is a equivalence:

 - `E : 𝓜(F(A)) ≌ 𝓡(A, R)`

The point I'm trying to make is that `restrict_scalars` is essentially `E` whereas true restriction of scalars is actually `G` (at least according to me).

Furthermore I claim that confusion about when to use `restrict_scalars` and when to use `is_scalar_tower` is the result of people invoking `E` because of the name of this type synonym, when what they really wanted was `G`.

ocfnash replied:
I've just removed this paragraph as I now think it's unlikely to help no matter what the form.

jcommelin replied:
@ocfnash Aha! Thanks a lot for your explanation. I agree that removing it is ok. But if you want to restore it in some form, then this last version with informal mathematical language is certainly the one that was most helpful to me.

ocfnash replied:
I'm now convinced removing the paragraph is the way to go. The quotient of the effort required to actually make this extra point clearly versus its actual value is extremely low!

I think the job of this comment is to say to the person: "are you sure you shouldn't be using `is_scalar_tower` instead?" and it does that.

ocfnash committed:
Add missing article

ocfnash committed:
Remove confusing paragraph

jcommelin commented on src/algebra/algebra/basic.lean:
```diff
@@ -1404,20 +1404,40 @@ end linear_map
-Warning: use this type synonym judiciously!
-The preferred way of working with an `A`-module `M` as `R`-module (where `A` is an `R`-algebra),
-is by `[module R M] [module A M] [is_scalar_tower R A M]`.
+/-- A type synonym representing the functor from the category of modules over `A`, regarded as a
```
This sentence still contains the (to me) confusing statement... I think it would already help if you replace `functor` be `equivalence of categories`. And also, talking about representations instead of modules, might help.

ocfnash replied:
Sorry, it was sloppy of me to forget about this opening remark. I've just pushed some new wording which hopefully helps.

I should say, I don't mind we give up on this PR: I'm not sure I'm helping!

ocfnash committed:
Remove confusing remark.

In particular the word `module` is no longer overloaded to refer both to a
Mathlib `module` and an algebra representation.

Vierkantor commented on src/algebra/algebra/basic.lean:
```diff
@@ -1430,18 +1451,27 @@ instance [I : add_comm_group M] : add_comm_group (restrict_scalars R A M) := I
-instance : module R (restrict_scalars R A M) :=
+/-- When `M` is a module over a semiring `A`, and `A` is an algebra over `R`, then `M` inherits a
+module structure over `R`. -/
+instance [comm_semiring R] [algebra R A] : module R (restrict_scalars R A M) :=
```
In the spirit of instances like `submodule.has_scalar`, if the following change is easy to apply, it might mean we only have to apply `restrict_scalars` once.
```suggestion
instance [comm_semiring R] [comm_semiring S] [algebra S R] [algebra R A] [algebra S A] [is_scalar_tower S R A] : 
  module S (restrict_scalars R A M) :=
```

ocfnash replied:
Hmm, I think it's not so easy to think of a situation where we'd use this but I guess it can't hurt if it works. I've just added a commit to see :-)

ocfnash replied:
This built successfully but raised a linter error. This is easily fixed but made me think about this a bit more. Having done so I now feel this is tilting from useful generalisation into non-trivial complication so I've reverted this change, especially since we don't have an application to hand.

Vierkantor reviewed (approved):
Thanks for the very extensive docstrings!

ocfnash committed:
CR suggestion: generalise `restrict_scalars.module`

I actually find it very hard to think of a situation where this would
be necessary but I guess it doesn't hurt.

ocfnash committed:
Revert "CR suggestion: generalise `restrict_scalars.module`"

This reverts commit 444b0b46b96416339f40c1b2e3172bfad12365b5.

Vierkantor reviewed (approved):
No worries, the extra `scalar_tower` would have been a nice bonus if it was easy enough to achieve. I'll let @jcommelin have the final say about getting this merged.

ocfnash committed:
Merge branch 'master' into restrict_scalars_equiv

jcommelin reviewed (approved):
Thanks :tada:

bors merge
